Specifications/Language/5_Grammar/QSharpLexer.g4

```diff
@@ -0,0 +1,190 @@
+lexer grammar QSharpLexer;
+
+// Keywords
+
+Adj : 'Adj';
+AdjointFunctor : 'Adjoint';
+AdjointGenerator : 'adjoint';
+And : 'and';
+Apply : 'apply';
+As : 'as';
+Auto : 'auto';
+BigInt : 'BigInt';
+Body : 'body';
+Bool : 'Bool';
+Borrowing : 'borrowing';
+ControlledFunctor : 'Controlled';
+ControlledGenerator : 'controlled';
+Ctl : 'Ctl';
+Distribute : 'distribute';
+Double : 'Double';
+Elif : 'elif';
+Else : 'else';
+Fail : 'fail';
+False : 'false';
+Fixup : 'fixup';
+For : 'for';
+Function : 'function';
+If : 'if';
+In : 'in';
+Int : 'Int';
+Internal : 'internal';
+Intrinsic : 'intrinsic';
+Invert : 'invert';
+Is : 'is';
+Let : 'let';
+Mutable : 'mutable';
+Namespace : 'namespace';
+New : 'new';
+Newtype : 'newtype';
+Not : 'not';
+One : 'One';
+Open : 'open';
+Operation : 'operation';
+Or : 'or';
+Pauli : 'Pauli';
+PauliI : 'PauliI';
+PauliX : 'PauliX';
+PauliY : 'PauliY';
+PauliZ : 'PauliZ';
+Qubit : 'Qubit';
+Range : 'Range';
+Repeat : 'repeat';
+Result : 'Result';
+Return : 'return';
+Self : 'self';
+Set : 'set';
+String : 'String';
+True : 'true';
+Unit : 'Unit';
+Until : 'until';
+Using : 'using';
+While : 'while';
+Within : 'within';
+Zero : 'Zero';
+
+// Operators
+
+AndEqual : 'and=';
+ArrowLeft : '<-';
+ArrowRight : '->';
+Asterisk : '*';
+AsteriskEqual : '*=';
+At : '@';
+Bang : '!';
+BraceLeft : '{' -> pushMode(DEFAULT_MODE);
+BraceRight : '}' { if (!_modeStack.isEmpty()) popMode(); };
+BracketLeft : '[';
+BracketRight : ']';
+Caret : '^';
+CaretEqual : '^=';
+Colon : ':';
+Comma : ',';
+DollarQuote : '$"' -> pushMode(INTERPOLATED);
+Dot : '.';
+DoubleColon : '::';
+DoubleDot : '..';
+DoubleEqual : '==';
+DoubleQuote : '"' -> pushMode(STRING);
+Ellipsis : '...';
+Equal : '=';
+FatArrowRight : '=>';
+Greater : '>';
+GreaterEqual : '>=';
+Less : '<';
+LessEqual : '<=';
+Minus : '-';
+MinusEqual : '-=';
+NotEqual : '!=';
+OrEqual : 'or=';
+ParenLeft : '(';
+ParenRight : ')';
+Percent : '%';
+PercentEqual : '%=';
+Pipe : '|';
+Plus : '+';
+PlusEqual : '+=';
+Question : '?';
+Semicolon : ';';
+Slash : '/';
+SlashEqual : '/=';
+TripleAmpersand : '&&&';
+TripleAmpersandEqual : '&&&=';
+TripleCaret : '^^^';
+TripleCaretEqual : '^^^=';
+TripleGreater : '>>>';
+TripleGreaterEqual : '>>>=';
+TripleLess : '<<<';
+TripleLessEqual : '<<<=';
+TriplePipe : '|||';
+TriplePipeEqual : '|||=';
+TripleTilde : '~~~';
+Underscore : '_';
+With : 'w/';
+WithEqual : 'w/=';
+
+// Literals
+
+fragment Digit : [0-9];
+
+IntegerLiteral
+    : Digit+
+    | ('0x' | '0X') [0-9a-fA-F]+
+    | ('0o' | '0O') [0-7]+
+    | ('0b' | '0B') [0-1]+
+    ;
+
+BigIntegerLiteral : IntegerLiteral ('L' | 'l');
+
+DoubleLiteral
+    : Digit+ '.' Digit+
+    | '.' Digit+
+    | Digit+ '.' { _input.LA(1) != '.' }?
```

Comment:
Hi Sarah, thanks for fixing this issue in the original repo. Isn't this semantic predicate language-specific (Java in this case)? We should perhaps add a note as we noticed in [the example you shared](https://github.com/Kotlin/kotlin-spec/blob/release/grammar/src/main/antlr/KotlinLexer.g4#L44).

Reply:
Yes, that's a good point. I added a note to the README instead of the source code since the grammar is actually using Java code in two places (there's an action in the parser as well).

Reply:
Thanks, like Bettina asks [below](https://github.com/microsoft/qsharp-language/pull/45#discussion_r512355667), I think it will be good to mention it in the grammar as well so it is clear what is part of the spec and what is implementation detail.

Reply:
The `-> pushMode(...)` and `-> popMode` things are actually built into ANTLR, so they're target-language-agnostic. The only other Java-specific code is

https://github.com/microsoft/qsharp-language/blob/1118ce6359d4509e9c3f6c766453bd019162354b/Specifications/Language/5_Grammar/QSharpLexer.g4#L76